spylib/token.py

```diff
@@ -0,0 +1,428 @@
+                'api_version': api_version,
+                'rest_bucket_max': rest_bucket_max,
+                'rest_leak_rate': rest_leak_rate,
+                'graphql_bucket_max': graphql_bucket_max,
+                'graphql_leak_rate': graphql_leak_rate,
+                'access_token_invalid': False,
+                'client': AsyncClient(),
+                'oauth_url': f'https://{store_name}.myshopify.com/admin/oauth/access_token',
+                'api_url': f'https://{store_name}.myshopify.com/admin/api/{api_version}',
+            }
+        )
+
+    @classmethod
+    async def new(
+        cls,
+        store_name: str,
+        client_id: str,
+        client_secret: str,
+        code: str,
+    ):
+        """
+        Creates a new instance of the object and requests the token from the API
+        as it requires an async call.
+        """
+        token = cls(store_name=store_name)
+        await token.reset_token(client_id, client_secret, code)
+        return token
+
+    @abstractmethod
+    async def save_token(self):
+        """
+        This method handles saving the token. By default this does nothing,
+        therefore the developer should override this.
+        """
+        raise NotImplementedError("Please Implement the save_token method")
+
+    @classmethod
+    @abstractmethod
+    def load_token(self):
+        """
+        This method handles loading the token. By default this does nothing,
+        therefore the developer should override this.
+        """
+        raise NotImplementedError("Please Implement the load_token method.")
+
+    async def _get_token(
+        self, client_id: str, client_secret: str, authorization_code: str
+    ) -> dict:
+        """
+        Makes a request to the access_token endpoint using the client secret,
+        client_id, and the code generated by the OAuth.
+        """
+
+        httpclient = AsyncClient()
+
+        jsondata = {
+            'client_id': client_id,
+            'client_secret': client_secret,
+            'code': authorization_code,
+        }
+
+        response = await httpclient.request(
+            method='post',
+            url=self.oauth_url,
+            json=jsondata,
+            timeout=20.0,
+        )
+        if response.status_code != 200:
+            message = (
+                f'Problem retrieving access token. Status code: {response.status_code} {jsondata}'
+                f'response=> {response.text}'
+            )
+            logger.error(message)
+            raise ValueError(message)
+
+        return response.json()
+
+    async def reset_token(self, client_id: str, client_secret: str, code: str):
+        """Use this function to initialize a new access token for this store"""
+        await self._get_token(client_id, client_secret, code)
+        self.access_token_invalid = False
+
+    @classmethod
+    def get_unique_id(cls) -> str:
+        return ShortUUID().random(length=10)
```

Comment:
Actually this should come from Toolip. This is a pure function. Nothing to do with this class.
Also a classmethod should return the class it belongs too, though not actually enforced. I've used them to return lists of the parent class ... that may not be pythonic but I did it anyway  :wink: